### ISSUE TITLE
SNAPCRAFT_PROJECT_VERSION is empty if not explicitly set at top level

### ISSUE THREAD

**Issue by Hook25** [snapcore/snapcraft#4231]
### Bug Description

If the version field is calculated via `adopt-info` instead of explicitly set in the top level version field, the `SNAPCRAFT_PROJECT_VERSION` envvar is empty during the recipe execution (both in `override-build` and in `build-environment`). This is the case even if the step that sets this information (via `snapcraft set-version`) is the first in the recipe.

This could either be the intended behavior of this variable, but that is unclear from the documentation and it should, in my opinion, yield a warning, or a bug. 

### To Reproduce

1. Create a recipe that sets the version at runtime
2. Use the `SNAPCRAFT_PROJECT_VERSION` anywhere in any subsequent step

**Expected:** The environment variable has the set value

**Result:** The envvar is empty

### Environment

N/A

### snapcraft.yaml

```shell
https://github.com/canonical/checkbox/blob/fix_snap_builds_scm/checkbox-core-snap/series22/snap/snapcraft.yaml
```

### Relevant log output

```shell
N/A
```

### Additional context

N/A

**Comment by sergiusens** [comment]
Thanks for the report.
Your snapcraft.yaml link leads to a 404
Where in the documentation is this not clear? We have no concept of dynamic environment variables, I can make that adjustment to the documentation if you tell me where it was unclear.

**Comment by Hook25** [comment]
Main snapcraft.yaml for series22 [link](https://github.com/canonical/checkbox/blob/main/checkbox-core-snap/series22/snap/snapcraft.yaml), I have linked a PR file above because at the time the change was not in main. 

[This forum page](https://forum.snapcraft.io/t/parts-environment-variables/12271) is the one that I have used a lot to remember the name of the variables available during builds. It states: 
> SNAPCRAFT_PROJECT_VERSION | the snapcraft project version as set by snapcraft.yaml

I think this should say something like 
> The snapcraft project version if it is **statically** set by snapcraft.yaml

If I can give my two cents on this, I would really like to see that variable being dynamically filled once the adopt-version part is run. I think it makes all builds that use part versions cleaner. If you look at our snapcraft.yaml, we keep reading the version file and using it to set the library version, not very pretty imo, it would be way clearer to the reader if we could write `SETUPTOOLS_SCM_PRETEND_VERSION : $SNAPCRAFT_PROJECT_VERSION` or something like that. Also, if that is not the case, I feel like the name of the variable is a bit misleading
